float64 lower
float64 upper
string pathname